File: jas_ocaml/lib/canvas/canvas_tool.ml
(** Tool protocol: shared types, geometry helpers, and tool factory.

    Tool implementations live in separate modules:
    - Selection_tool: selection, group selection, direct selection
    - Drawing_tool: line, rect, polygon
    - Pen_tool: pen (Bezier path creation)
    - Text_tool: text placement and editing *)

(* ------------------------------------------------------------------ *)
(* Tool context                                                        *)
(* ------------------------------------------------------------------ *)

type tool_context = {
  model : Model.model;
  controller : Controller.controller;
  hit_test_selection : float -> float -> bool;
  hit_test_handle : float -> float -> (int list * int * string) option;
  hit_test_text : float -> float -> (int list * Element.element) option;
  hit_test_path_curve : float -> float -> (int list * Element.element) option;
  request_update : unit -> unit;
  start_text_edit : int list -> Element.element -> unit;
  commit_text_edit : unit -> unit;
  draw_element_overlay : Cairo.context -> Element.element -> int list -> unit;
}

(* ------------------------------------------------------------------ *)
(* Tool class type                                                     *)
(* ------------------------------------------------------------------ *)

class type canvas_tool = object
  method on_press : tool_context -> float -> float -> shift:bool -> alt:bool -> unit
  method on_move : tool_context -> float -> float -> shift:bool -> dragging:bool -> unit
  method on_release : tool_context -> float -> float -> shift:bool -> alt:bool -> unit
  method on_double_click : tool_context -> float -> float -> unit
  method on_key : tool_context -> int -> bool
  method on_key_release : tool_context -> int -> bool
  method draw_overlay : tool_context -> Cairo.context -> unit
  method activate : tool_context -> unit
  method deactivate : tool_context -> unit
end

(* ------------------------------------------------------------------ *)
(* Geometry helpers                                                    *)
(* ------------------------------------------------------------------ *)

let constrain_angle sx sy ex ey =
  let dx = ex -. sx and dy = ey -. sy in
  let dist = sqrt (dx *. dx +. dy *. dy) in
  if dist = 0.0 then (ex, ey)
  else
    let angle = atan2 dy dx in
    let snapped = Float.round (angle /. (Float.pi /. 4.0)) *. (Float.pi /. 4.0) in
    (sx +. dist *. cos snapped, sy +. dist *. sin snapped)

(* Shared tool constants *)
let hit_radius = 8.0          (* pixels to detect a click on a control point *)
let handle_draw_size = 10.0   (* diameter of control-point handles in pixels *)
let drag_threshold = 4.0      (* pixels of movement before a click becomes a drag *)
let paste_offset = 24.0       (* translation in pt applied when pasting *)
let long_press_ms = 500       (* milliseconds before a press becomes a long-press *)
let polygon_sides = 5         (* default number of sides for the polygon tool *)

let regular_polygon_points x1 y1 x2 y2 n =
  let ex = x2 -. x1 and ey = y2 -. y1 in
  let s = sqrt (ex *. ex +. ey *. ey) in
  if s = 0.0 then List.init n (fun _ -> (x1, y1))
  else
    let mx = (x1 +. x2) /. 2.0 and my = (y1 +. y2) /. 2.0 in
    let px = -. ey /. s and py = ex /. s in
    let d = s /. (2.0 *. tan (Float.pi /. float_of_int n)) in
    let cx = mx +. d *. px and cy = my +. d *. py in
    let r = s /. (2.0 *. sin (Float.pi /. float_of_int n)) in
    let theta0 = atan2 (y1 -. cy) (x1 -. cx) in
    List.init n (fun k ->
      let angle = theta0 +. 2.0 *. Float.pi *. float_of_int k /. float_of_int n in
      (cx +. r *. cos angle, cy +. r *. sin angle))

(* ------------------------------------------------------------------ *)
(* Shared defaults                                                     *)
(* ------------------------------------------------------------------ *)

let default_stroke = Some Element.{
  stroke_color = { r = 0.0; g = 0.0; b = 0.0; a = 1.0 };
  stroke_width = 1.0;
  stroke_linecap = Butt;
  stroke_linejoin = Miter;
}

File: JasSwift/Sources/Canvas/CanvasTool.swift
import AppKit
import Foundation

// MARK: - Tool context

/// Facade passed to tools giving access to model, controller, and canvas services.
class ToolContext {
    let model: Model
    let controller: Controller
    let hitTestSelection: (NSPoint) -> Bool
    let hitTestHandle: (NSPoint) -> (path: ElementPath, anchorIdx: Int, handleType: String)?
    let hitTestText: (NSPoint) -> (ElementPath, Text)?
    let hitTestPathCurve: (Double, Double) -> (ElementPath, Element)?
    let requestUpdate: () -> Void
    let drawElementOverlayFn: (CGContext, Element, SelectionKind) -> Void

    init(model: Model,
         controller: Controller,
         hitTestSelection: @escaping (NSPoint) -> Bool,
         hitTestHandle: @escaping (NSPoint) -> (path: ElementPath, anchorIdx: Int, handleType: String)?,
         hitTestText: @escaping (NSPoint) -> (ElementPath, Text)?,
         hitTestPathCurve: @escaping (Double, Double) -> (ElementPath, Element)?,
         requestUpdate: @escaping () -> Void,
         drawElementOverlay: @escaping (CGContext, Element, SelectionKind) -> Void) {
        self.model = model
        self.controller = controller
        self.hitTestSelection = hitTestSelection
        self.hitTestHandle = hitTestHandle
        self.hitTestText = hitTestText
        self.hitTestPathCurve = hitTestPathCurve
        self.requestUpdate = requestUpdate
        self.drawElementOverlayFn = drawElementOverlay
    }

    var document: Document { model.document }

    func snapshot() { model.snapshot() }
}

// MARK: - Keyboard modifiers

/// Modifier keys passed to `onKeyEvent`. `cmd` is true if the platform's
/// command key (NSEvent.modifierFlags.command on macOS) is held.
public struct KeyMods {
    public var shift: Bool
    public var ctrl: Bool
    public var alt: Bool
    public var cmd: Bool
    public init(shift: Bool = false, ctrl: Bool = false, alt: Bool = false, cmd: Bool = false) {
        self.shift = shift; self.ctrl = ctrl; self.alt = alt; self.cmd = cmd
    }
}

// MARK: - CanvasTool protocol

/// Interface for canvas interaction tools.
protocol CanvasTool: AnyObject {
    func onPress(_ ctx: ToolContext, x: Double, y: Double, shift: Bool, alt: Bool)
    func onMove(_ ctx: ToolContext, x: Double, y: Double, shift: Bool, dragging: Bool)
    func onRelease(_ ctx: ToolContext, x: Double, y: Double, shift: Bool, alt: Bool)
    func onDoubleClick(_ ctx: ToolContext, x: Double, y: Double)
    func onKey(_ ctx: ToolContext, keyCode: UInt16) -> Bool
    func onKeyUp(_ ctx: ToolContext, keyCode: UInt16) -> Bool
    func drawOverlay(_ ctx: ToolContext, _ cgCtx: CGContext)
    func activate(_ ctx: ToolContext)
    func deactivate(_ ctx: ToolContext)

    // Optional in-place text editing surface (default no-op for most tools).
    func capturesKeyboard() -> Bool
    func isEditing() -> Bool
    func cursorOverride() -> String?
    func pasteText(_ ctx: ToolContext, _ text: String) -> Bool
    func onKeyEvent(_ ctx: ToolContext, _ key: String, _ mods: KeyMods) -> Bool
}

/// Default implementations for optional protocol methods.
extension CanvasTool {
    func onDoubleClick(_ ctx: ToolContext, x: Double, y: Double) {}
    func onKey(_ ctx: ToolContext, keyCode: UInt16) -> Bool { false }
    func onKeyUp(_ ctx: ToolContext, keyCode: UInt16) -> Bool { false }
    func activate(_ ctx: ToolContext) {}
    func deactivate(_ ctx: ToolContext) {}
    func capturesKeyboard() -> Bool { false }
    func isEditing() -> Bool { false }
    func cursorOverride() -> String? { nil }
    func pasteText(_ ctx: ToolContext, _ text: String) -> Bool { false }
    func onKeyEvent(_ ctx: ToolContext, _ key: String, _ mods: KeyMods) -> Bool { false }
}

// MARK: - Helpers

// MARK: - Shared tool constants

let hitRadius: CGFloat = 8.0        // pixels to detect a click on a control point
let handleDrawSize: CGFloat = 10.0  // diameter of control-point handles in pixels
let dragThreshold: Double = 4.0     // pixels of movement before a click becomes a drag
let pasteOffset: Double = 24.0      // translation in pt applied when pasting
let longPressDuration: Double = 0.5 // seconds before a press becomes a long-press
let polygonSides = 5                // default number of sides for the polygon tool
let flattenSteps = elementFlattenSteps  // shared constant from Element.swift

let toolSelectionColor = CGColor(red: 0, green: 0.47, blue: 1.0, alpha: 1.0)

func constrainAngle(_ sx: Double, _ sy: Double, _ ex: Double, _ ey: Double) -> (Double, Double) {
    let dx = ex - sx, dy = ey - sy
    let dist = hypot(dx, dy)
    guard dist > 0 else { return (ex, ey) }
    let angle = atan2(dy, dx)
    let snapped = (angle / (.pi / 4)).rounded() * (.pi / 4)
    return (sx + dist * cos(snapped), sy + dist * sin(snapped))
}

// regularPolygonPoints lives in Geometry/RegularShapes.swift.

// MARK: - Tool registry

/// Look up a tool spec inside an already-loaded workspace and build
/// a YamlTool from it. Returns nil only when the spec is missing or
/// malformed — workspace failure is reported by [loadWorkspaceTools]
/// up-front, not per call. The dispatcher in CanvasSubwindow keys
/// the tool dict with optional chaining (tools[currentTool]?), so a
/// missing tool downgrades to a no-op rather than a crash.
func loadYamlTool(_ id: String, in ws: WorkspaceData) -> YamlTool? {
    guard let tools = ws.data["tools"] as? [String: Any],
          let spec = tools[id] as? [String: Any] else {
        return nil
    }
    return YamlTool.fromWorkspaceTool(spec)
}

/// Load workspace tools and assemble the registry. Tiered failure
/// modes:
///   * Workspace nil → fatalError. The whole app cannot run without
///     workspace.json; failing fast is correct.
///   * Per-tool nil → log + omit. Other tools still load, the omitted
///     tool just won't activate when chosen. This is the right
///     trade-off because adding a new tool spec shouldn't be able to
///     brick the app.
/// Native-only tools (Type / TypeOnPath per NATIVE_BOUNDARY.md §6)
/// are added unconditionally.
func createTools() -> [Tool: CanvasTool] {
    guard let ws = WorkspaceData.load() else {
        fatalError("workspace/workspace.json missing or malformed — cannot run without it")
    }
    let yamlTools: [(Tool, String)] = [
        (.selection,           "selection"),
        (.partialSelection,    "partial_selection"),
        (.interiorSelection,   "interior_selection"),
        (.pen,                 "pen"),
        (.addAnchorPoint,      "add_anchor_point"),
        (.deleteAnchorPoint,   "delete_anchor_point"),
        (.anchorPoint,         "anchor_point"),
        (.pencil,              "pencil"),
        (.pathEraser,          "path_eraser"),
        (.smooth,              "smooth"),
        (.line,                "line"),
        (.rect,                "rect"),
        (.roundedRect,         "rounded_rect"),
        (.polygon,             "polygon"),
        (.star,                "star"),
        (.lasso,               "lasso"),
    ]
    var registry: [Tool: CanvasTool] = [
        .typeTool:   TypeTool(),
        .typeOnPath: TypeOnPathTool(),
    ]
    for (kind, id) in yamlTools {
        if let tool = loadYamlTool(id, in: ws) {
            registry[kind] = tool
        } else {
            NSLog("[Jas] tool \"%@\" not loaded — workspace.json missing the spec", id)
        }
    }
    return registry
}
